### ISSUE TITLE
Support ES256, ES384, and ES512

### ISSUE THREAD

**Issue by jonblatho** [IBM-Swift/Swift-JWT#24]
Long shot here, but what are the odds that this can support ECDSA signatures without depending on Apple platforms? This is important for many applications which need to interact with Apple's cloud services (i.e., APNS, MusicKit, etc.).

In investigating this issue myself, it looks like it's at least *somewhat* straightforward to accomplish this using Apple's Security, but Security obviously isn't available on Linux. Crypto stuff is way over my head and I'm worried that I can't implement it properly working alone, and it seems like there's no Swift library out there yet that handles ECDSA signing without depending on Security. Without that, any such solution is dead on arrival for many server-side use cases which need to run on Linux.

There does exist a [crypto library](https://github.com/alexaubry/ServerCrypto) which claims to handle ECDSA signing, but in my experience it doesn't work as expected.

So, how hard would it be to pull this off? I'd be happy to help where I can, even if it's just contributing to inline docs since as I mentioned, proper implementation is important of course, and I'm more than a little worried that I can't do it right by myself.

**Comment by Andrew-Lees11** [comment]
Hello @jonblatho 
I'll have a look into supporting ECDSA and how difficult it would be to implement. We currently use [BlueCryptor](https://github.com/IBM-Swift/BlueCryptor) for cross platform Security so I'll see if we can use that.

**Comment by Andrew-Lees11** [comment]
So [BlueCryptor](https://github.com/IBM-Swift/BlueCryptor) doesn't currently have support for ECDSA. Since we would like to use ECDSA to sign and verify JWT's for ES256, we are going to add ecliptic curve support. I think the most logical place for this to go would be in BlueCryptor as opposed to creating a new repo.
For iOS/MacOS, we can use CommonCrypto and the Apple security API's documented [here](https://developer.apple.com/documentation/security/certificate_key_and_trust_services/keys/signing_and_verifying). [CupertinoJWT](https://github.com/ethanhuang13/CupertinoJWT) already allows signing of JWT's using this implementation style.
For linux we would create a swift wrapper for [<openssl/ecdsa.h>](https://www.openssl.org/docs/man1.0.2/crypto/ecdsa.html). 
We would then select the correct implementation based on your operating system and expose this as a common API. We would then be able to use this API to support ES256/ES384/ES512 in a similar way to how SwiftJWT supports RS256/ES384/ES512.
The majority of this work will be in BlueCryptor so I will raise in issue to track it on that repo.

**Comment by Andrew-Lees11** [comment]
Hey @jonblatho 
So I have a prototype for ES256 on [this branch](https://github.com/IBM-Swift/Swift-JWT/tree/issue.ECDSA).  It is working on MacOS and linux 16.04 without any extra dependencies. 

It signs and verifies using PEM encoded EC keys that can be [generated using OpenSSL](https://wiki.openssl.org/index.php/Command_Line_Elliptic_Curve_Operations). linux 18.04 is currently failing  so I will look into that as well as adding ES384 and ES512 support. 

Would you be able to help test this our and see if it works for you requirements?

**Comment by jonblatho** [comment]
Amazing! I’ll test it out over the weekend and let you know how it goes!

**Comment by jonblatho** [comment]
It works on my end on macOS!

As some guidance for others who might stumble upon this, since it wasn't immediately clear in the documentation, the OpenSSL command I used to go from Apple's .p8 file to the desired .pem file is as follows:

    $ openssl ec -in AuthKey_XXXXXXXXXX.p8 -out key.pem

Thank you for your work on this, @Andrew-Lees11! You've saved a lot of people a lot of headaches.

**Comment by Andrew-Lees11** [comment]
Closing now ES256, ES384 and ES512 have been added in the 3.3.0 release of SwiftJWT.
They should be able to take keys in either pem or p8 format.

**Comment by IuriiIaremenko** [comment]
@Andrew-Lees11 it looks like ES256 is not supported by Linux, only Apple platforms have es256 signer.

```swift
    /// Initialize a JWTSigner using the ECDSA SHA256 algorithm and the provided privateKey.
    /// - Parameter privateKey: The UTF8 encoded PEM private key, with either a "BEGIN EC PRIVATE KEY" or "BEGIN PRIVATE KEY" header.
    @available(OSX 10.13, iOS 11, tvOS 11.0, watchOS 4.0, *)
    public static func es256(privateKey: Data) -> JWTSigner {
        return JWTSigner(name: "ES256", signerAlgorithm: BlueECSigner(key: privateKey, curve: .prime256v1))
    }
```
